server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java

```diff
@@ -224,37 +247,97 @@ public synchronized void shutdownNode() {
         url.append("/extra/");
         url.append(NodeForceRestartServlet.class.getSimpleName());
         HttpPost post = new HttpPost(url.toString());
+        int responseStatusCode = HttpStatus.SC_NOT_FOUND;
         try {
-            client.execute(post);
-            appendMsgToCustomLog("Node " + machine + " shut-down successfully.");
+            HttpResponse response = client.execute(post);
+            responseStatusCode = response.getStatusLine().getStatusCode();
         } catch (IOException e) {
             LOGGER.log(Level.SEVERE, e.getMessage(), e);
         } finally {
+            if (responseStatusCode == HttpStatus.SC_OK) {
+                proxyLogger.info("Node " + machine + " shutdown successfully.");
+            } else {
+                proxyLogger.info("Node " + machine + " did not shutdown. Return code was " + responseStatusCode);
+            }
+
             try {
                 client.close();
             } catch (IOException e) {
                 LOGGER.log(Level.SEVERE, e.getMessage(), e);
             }
         }
+        stopNodeRecycleThread();
+        // remove the node from the hub registry, if it reported 200 OK for NodeForceRestartServlet request
+        if (responseStatusCode == HttpStatus.SC_OK) {
+            getRegistry().removeIfPresent(this);
+        }
+        LOGGER.exiting();
     }
 
     /**
      * Thread will recycle the node when all active sessions are completed
      */
     class NodeRecycleThread extends Thread {
+        private volatile boolean running;
+
+        NodeRecycleThread() {
+            running = false;
+        }
 
         @Override
         public void run() {
-            appendMsgToCustomLog("Started NodeRecycleThread to restart the node once all the sessions are completed.");
-            while (getTotalUsed() > 0) {
+            LOGGER.entering();
+            running = true;
+            int timeout = -1; // Default. Wait forever.
+            int expired = 0;
+
+            try {
+                timeout = ConfigParser.parse().getInt("nodeRecycleThreadWaitTimeout");
+            } catch (ConfigParserException e) { // NOSONAR
+                // ignore we have a default already
+            }
+
+            proxyLogger.fine("Started NodeRecycleThread with " + ((timeout == -1) ? "no" : "a " + timeout + " second")
+                    + " timeout.");
+            while (keepLooping(expired, timeout)) {
```

Comment:
Can we simplify the timeout loop here in run? also would a timeout of 0 be better to indicate "loop forever" (i.e no timeout) instead of "-1"